os/arch/arm/src/armv7-a/arm_mmu.c

```diff
@@ -254,3 +289,264 @@ void mmu_invalidate_region(uint32_t vstart, size_t size)
+{
+	uint32_t idx;
+	uint32_t val;
+	uint32_t end = start + size;
+	irqstate_t flags;
+
+	lldbg("start = 0x%08x end = 0x%08x size = %x\n", start, end, size);
+
+	// Run a loop until the entire region is mapped.
+	while (start < end) {
+		// Check if this address can be mapped to a section.
+		if (!(start & SECTION_MASK) && !(size & SECTION_MASK)) {
+			// Yes. Update the section entry in the the L1 page table.
+			idx = start >> 20;
+			val = start & PMD_PTE_PADDR_MASK;
+			mmu_set_flags(&val, ro, exec, true, global);
+
+			lldbg("Add section for addr 0x%08x idx = %d\n", start, idx);
+
+			if (global) {
+  				// If this update is for the common binary, then it is done
+				// in the kernel page tables and so the cache and tlbs need
+				// to be flushed and invalidated.
+				flags = enter_critical_section();
+				l1_pgtbl[idx] = val;
+				cp15_clean_dcache_bymva((uint32_t)&l1_pgtbl[idx]);
+  				mmu_invalidate_region(start, SECTION_SIZE);
+				leave_critical_section(flags);
+			} else {
+				l1_pgtbl[idx] = val;
+				cp15_clean_dcache_bymva((uint32_t)&l1_pgtbl[idx]);
+			}
+
+			// Advance the memory region address.
+			start += SECTION_SIZE;
+		} else {	// Check if this address can be mapped to a small page.
+
+			// Check if L2 page table is not created.
+			idx = (start & 0xfff00000) >> 20;
+			int l2_idx = 0;
+			uint32_t *l2_pgtbl = (uint32_t *)(l1_pgtbl[idx] & PMD_PTE_PADDR_MASK);
+			if ((l1_pgtbl[idx] & PMD_TYPE_MASK) != PMD_TYPE_PTE) {
+				// Yes. Allocate L2 page table for app.
+				l2_pgtbl = mmu_allocate_app_l2_pgtbl(app_id, l2_idx++);
+
+				lldbg("Allocated L2 pgtbl at 0x%08x\n", l2_pgtbl);
+				if (global) {
+  					// If this update is for the common binary, then it is done
+					// in the kernel page tables and so the cache and tlbs need
+					// to be flushed and invalidated.
+
+					flags = enter_critical_section();
+
+					// Fill default entries into L2 page table.
+					uint32_t tmp = start;
+					for (idx = 0; idx < L2_PGTBL_NENTRIES; idx++) {
+						val = tmp & PTE_SMALL_PADDR_MASK;
+						val |= MMU_MEMFLAGS;
+						l2_pgtbl[idx] = val;
+						cp15_clean_dcache_bymva((uint32_t)&l2_pgtbl[idx]);
+						tmp += 4096;
+					}
+
+					// Update L2 page table address in L1 page table. 
+					val = (uint32_t)l2_pgtbl & PMD_PTE_PADDR_MASK;
+					val |= MMU_L1_DATAFLAGS;
+					l1_pgtbl[idx] = val;
+
+					cp15_clean_dcache_bymva((uint32_t)&l1_pgtbl[idx]);
+					cp15_invalidate_tlb_bymva(start);
+					leave_critical_section(flags);
+				} else {
+					// Update L2 page table address in L1 page table.
+					val = (uint32_t)l2_pgtbl & PMD_PTE_PADDR_MASK;
+					val |= MMU_L1_PGTABFLAGS;
+					l1_pgtbl[idx] = val;
+					// dbg("Set l1 pte at 0x%08x = 0x%08x\n", &l1_pgtbl[idx], val);
+					cp15_clean_dcache_bymva((uint32_t)&l1_pgtbl[idx]);
+				}
+			}
+
+			// Update the L2 page table entry.
+  			idx = (start & 0x000ff000) >> 12;
+			val = start & PTE_SMALL_PADDR_MASK;
```

Comment:
If we are filling the addresses from only the start in the page table, it will miss some entries before start and after end. I think it might not be right?? (May cause prefetch abort when fetching the address that does not have an entry in secondary page table?)

Reply:
Agreed. I will change it. But it will not affect our current issue.